File: apps/ios/InnoLive/InnoLive/Features/OnDevicePrivacy/PrivacyMaskStabilizer.swift
#if DEBUG
import Foundation
import CoreGraphics

/// Aligns recent per-object masks to current detection boxes. New protection is immediate;
/// only removal of an old edge fades. Matching never grants an identity/whitelist exception.
nonisolated final class PrivacyMaskStabilizer {
    private var previous: [PrivacySegmentation.InstanceMask] = []
    private var previousTime: TimeInterval?
    private let fadeSeconds: Double = 0.12
    private let maximumGap: Double = 0.20

    func reset() {
        previous = []
        previousTime = nil
    }

    func apply(_ current: [PrivacySegmentation.InstanceMask], timestamp: TimeInterval) throws -> [UInt8] {
        let size = PrivacySegmentation.maskSize
        guard timestamp.isFinite, current.allSatisfy({ $0.bytes.count == size * size }) else {
            throw PrivacyModelError.outputContract
        }
        let interval = previousTime.map { timestamp - $0 } ?? maximumGap
        if interval <= 0 || interval >= maximumGap { previous = [] }
        let decay = Float(exp(-max(0, interval) / fadeSeconds))
        var available = Set(previous.indices)
        var stabilized: [PrivacySegmentation.InstanceMask] = []
        for instance in current {
            let box = instance.detection.box
            let match = available
                .filter { previous[$0].detection.classID == instance.detection.classID }
                .map { ($0, PrivacySegmentation.iou(box, previous[$0].detection.box)) }
                .filter { $0.1 >= 0.30 }
                .max { $0.1 < $1.1 }?.0
            guard let match else {
                stabilized.append(instance)
                continue
            }
            available.remove(match)
            let older = previous[match]
            let scale = CGFloat(size) / CGFloat(PrivacySegmentation.inputSize)
            let destination = box.applying(CGAffineTransform(scaleX: scale, y: scale))
            let source = older.detection.box.applying(CGAffineTransform(scaleX: scale, y: scale))
            guard destination.width > 0, destination.height > 0 else {
                throw PrivacyModelError.outputContract
            }
            let x0 = max(0, Int(floor(destination.minX)))
            let x1 = min(size, Int(ceil(destination.maxX)))
            let y0 = max(0, Int(floor(destination.minY)))
            let y1 = min(size, Int(ceil(destination.maxY)))
            var bytes = instance.bytes
            if x0 < x1 && y0 < y1 {
                for y in y0..<y1 {
                    let oldY = source.minY + (CGFloat(y) + 0.5 - destination.minY) / destination.height * source.height - 0.5
                    for x in x0..<x1 {
                        let oldX = source.minX + (CGFloat(x) + 0.5 - destination.minX) / destination.width * source.width - 0.5
                        let faded = Self.sample(older.bytes, x: oldX, y: oldY, size: size) * decay
                        let index = y * size + x
                        bytes[index] = max(bytes[index], UInt8(clamping: Int(faded)))
                    }
                }
            }
            stabilized.append(.init(detection: instance.detection, bytes: bytes))
        }
        // Unmatched/disappeared detections are not painted at stale screen coordinates.
        previous = stabilized
        previousTime = timestamp
        return PrivacySegmentation.union(stabilized)
    }

    private static func sample(_ bytes: [UInt8], x: CGFloat, y: CGFloat, size: Int) -> Float {
        guard x >= 0, y >= 0, x <= CGFloat(size - 1), y <= CGFloat(size - 1) else { return 0 }
        let x0 = Int(floor(x)), y0 = Int(floor(y))
        let x1 = min(x0 + 1, size - 1), y1 = min(y0 + 1, size - 1)
        let dx = Float(x - CGFloat(x0)), dy = Float(y - CGFloat(y0))
        let top = Float(bytes[y0 * size + x0]) * (1 - dx) + Float(bytes[y0 * size + x1]) * dx
        let bottom = Float(bytes[y1 * size + x0]) * (1 - dx) + Float(bytes[y1 * size + x1]) * dx
        return top * (1 - dy) + bottom * dy
    }
}
#endif

File: apps/android/InnoLive/app/src/main/java/com/framework/innolive/feature/live/privacy/PrivacyMaskStabilizer.kt
package com.framework.innolive.feature.live.privacy

import kotlin.math.ceil
import kotlin.math.exp
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min

/** Keeps a matched object's older edge briefly, aligned to its current box. */
internal class PrivacyMaskStabilizer {
    private var previous = emptyList<PrivacySegmentation.InstanceMask>()
    private var previousTimeSeconds: Double? = null

    fun reset() {
        previous = emptyList()
        previousTimeSeconds = null
    }

    fun apply(current: List<PrivacySegmentation.InstanceMask>, atSeconds: Double): ByteArray {
        val size = PrivacySegmentation.MASK_SIZE
        require(atSeconds.isFinite() && current.all { it.bytes.size == size * size })
        val interval = previousTimeSeconds?.let { atSeconds - it } ?: MAXIMUM_GAP_SECONDS
        if (interval <= 0 || interval >= MAXIMUM_GAP_SECONDS) previous = emptyList()
        val decay = exp(-max(0.0, interval) / FADE_SECONDS).toFloat()
        val available = previous.indices.toMutableSet()
        val stabilized = current.map { instance ->
            val match = available.filter { previous[it].detection.classId == instance.detection.classId }
                .map { it to previous[it].detection.box.intersects(instance.detection.box) }
                .filter { it.second >= .30f }
                .maxByOrNull { it.second }?.first
                ?: return@map instance
            available.remove(match)
            val older = previous[match]
            val destination = instance.detection.box
            val source = older.detection.box
            require(destination.width > 0 && destination.height > 0)
            val scale = size.toFloat() / PrivacySegmentation.INPUT_SIZE
            val left = destination.left * scale
            val top = destination.top * scale
            val right = destination.right * scale
            val bottom = destination.bottom * scale
            val sourceLeft = source.left * scale
            val sourceTop = source.top * scale
            val sourceWidth = source.width * scale
            val sourceHeight = source.height * scale
            val x0 = max(0, floor(left).toInt())
            val x1 = min(size, ceil(right).toInt())
            val y0 = max(0, floor(top).toInt())
            val y1 = min(size, ceil(bottom).toInt())
            val bytes = instance.bytes.copyOf()
            for (y in y0 until y1) {
                val oldY = sourceTop + (y + .5f - top) / (bottom - top) * sourceHeight - .5f
                for (x in x0 until x1) {
                    val oldX = sourceLeft + (x + .5f - left) / (right - left) * sourceWidth - .5f
                    val faded = (sample(older.bytes, oldX, oldY, size) * decay).toInt().coerceIn(0, 255)
                    val offset = y * size + x
                    if (faded > (bytes[offset].toInt() and 255)) bytes[offset] = faded.toByte()
                }
            }
            PrivacySegmentation.InstanceMask(instance.detection, bytes)
        }
        previous = stabilized
        previousTimeSeconds = atSeconds
        return PrivacySegmentation.union(stabilized)
    }

    private fun sample(bytes: ByteArray, x: Float, y: Float, size: Int): Float {
        if (x < 0 || y < 0 || x > size - 1 || y > size - 1) return 0f
        val x0 = floor(x).toInt()
        val y0 = floor(y).toInt()
        val x1 = min(x0 + 1, size - 1)
        val y1 = min(y0 + 1, size - 1)
        val dx = x - x0
        val dy = y - y0
        val top = (bytes[y0 * size + x0].toInt() and 255) * (1 - dx) +
            (bytes[y0 * size + x1].toInt() and 255) * dx
        val bottom = (bytes[y1 * size + x0].toInt() and 255) * (1 - dx) +
            (bytes[y1 * size + x1].toInt() and 255) * dx
        return top * (1 - dy) + bottom * dy
    }

    companion object {
        private const val FADE_SECONDS = .12
        private const val MAXIMUM_GAP_SECONDS = .20
    }
}
